akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Demux.scala

```diff
@@ -357,25 +367,64 @@ private[http2] abstract class Http2Demux(http2Settings: Http2CommonSettings, ini
         }
       })
 
+      // -----------------------------------------------------------------
       // FIXME: What if user handler doesn't pull in new substreams? Should we reject them
-      //        after a while or buffer only a limited amount? We should also be able to
-      //        keep the buffer limited to the number of concurrent streams as negotiated
-      //        with the other side.
+      //        after a while or buffer only a limited amount?
       val bufferedSubStreamOutput = new BufferedOutlet[Http2SubStream](substreamOut)
       override def dispatchSubstream(initialHeaders: ParsedHeadersFrame, data: Source[Any, Any], correlationAttributes: Map[AttributeKey[_], _]): Unit =
         bufferedSubStreamOutput.push(Http2SubStream(initialHeaders, data, correlationAttributes))
 
+      // -----------------------------------------------------------------
+      private var completing = false
+      private def onTryComplete(): Unit = {
+        completing = true
+        if (safeToComplete) complete()
+      }
+      override def onAllStreamsClosed(): Unit = if (safeToComplete) complete()
+      override def onAllDataFlushed(): Unit = if (safeToComplete) complete()
+
+      private def safeToComplete = {
+        if (isServer) {
+          completing && activeStreamCount() == 0 && (!isClosed(frameOut) && multiplexer.hasFlushedAllData)
+        } else {
+          completing && activeStreamCount() == 0
+        }
+      }
+
+      // Customized replacement for completeStage()
+      private def complete(): Unit = {
+        cancel(substreamIn)
+        cancel(frameIn)
+        complete(frameOut)
+        // Unlike the default `completeStage`, we need to delegate to
+        // bufferedSubStreamOutput#complete(substreamOut) instead of complete(substreamOut)
+        bufferedSubStreamOutput.complete()
+      }
+
+      // -----------------------------------------------------------------
       setHandler(substreamIn, new InHandler {
         def onPush(): Unit = {
           val sub = grab(substreamIn)
           handleOutgoingCreated(sub)
           // Once the incoming stream is handled, we decide if we need to pull more.
           tryPullSubStreams()
         }
+
+        override def onUpstreamFinish(): Unit = {
+          // marks StreamHandling as ready for completion
+          onTryComplete()
+          scheduleOnce(CompletionTimeout, settings.completionTimeout)
+        }
+
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          onTryComplete()
```

Comment:
This converts the failure to regular completion when the completion condition holds. Is that what we want?

Reply:
That's a leftover.